### ISSUE TITLE
Polymorphism requires turning strictMode off

### ISSUE THREAD

**Issue by hshar7** [Kotlin/kotlinx.serialization#700]
**Describe the bug**
`type` is bugging out as `Strict JSON encountered unknown key type` when parsing json for a polymorphic type.

**To Reproduce**
Attach a code snippet or test data if possible.
```
import kotlinx.serialization.SerialName
import kotlinx.serialization.json.Json
import kotlinx.serialization.Serializable
import kotlinx.serialization.UnstableDefault
import kotlinx.serialization.json.JsonConfiguration

@Serializable
data class EventeumPayload(
        val id: String,
        val type: String,
        val details: EventDetails,
        val retries: Int
) {
    companion object {
        @UnstableDefault
        fun fromJson(data: String): EventeumPayload {
            return Json(JsonConfiguration(strictMode = false)).parse(serializer(), data)
        }
    }
}

@Serializable
data class EventDetails(
        val nonIndexedParameters: List<EventValueType>,
)

@Serializable
sealed class EventValueType {
    @Serializable
    @SerialName("string")
    public data class EventValueTypeString(val value: String) : EventValueType()

    @Serializable
    @SerialName("address")
    public data class EventValueTypeAddress(val value: String) : EventValueType()

    @Serializable
    @SerialName("uint256")
    public data class EventValueTypeInt(val value: Int) : EventValueType()

    @Serializable
    @SerialName("bool")
    public data class EventValueTypeBool(val value: Int) : EventValueType()

    @Serializable
    @SerialName("uint256[]")
    public data class EventValueTypeObject(val value: List<EventValueType>) : EventValueType()
}
```

**Expected behavior**
`return Json.parse(serializer(), data)` should work

**Environment**
 - Kotlin version: 1.3.30
 - Library version: 0.14.0
 - Kotlin platforms: JVM
 - Gradle version: Maven

**Comment by qwwdfsad** [comment]
Could you please attach a test data that parser fails to read?
I've tried to do the following:
```
val data = EventeumPayload(
    "a", "b", EventDetails(
        listOf(
            EventValueType.EventValueTypeString("a"),
            EventValueType.EventValueTypeObject(listOf(EventValueType.EventValueTypeInt(1)))
        )
    ), 3
)
val str = Json(JsonConfiguration.Default).stringify(data)
println(str)
println(EventeumPayload.fromJson(str))
```

and it seems to work

**Comment by hshar7** [comment]
@qwwdfsad 
Here's the data that I was parsing:
```
{
    "id": "0x719dc02d633a4693ebffde7fbf29acbad98207a3430f6e4234d000a390f11a45-0x84ad721ff827e0e0abb4e55d5dac39631634e777fa90b30a75d7627004163501-39",
    "type": "CONTRACT_EVENT",
    "details": {
        "name": "TournamentIssued",
        "filterId": "tournamentIssuedEvent",
        "nodeName": "default",
        "indexedParameters": [],
        "nonIndexedParameters": [
            {
                "type": "uint256",
                "value": 1
            },
            {
                "type": "address",
                "value": "0xB6E58769550608DEF3043DCcbBE1Fa653af23151"
            },
            {
                "type": "string",
                "value": "QmbTRqZPv5jYTMFkWjog3ZbsrwwMRqJzd5WCqvkCCk1E7z"
            },
            {
                "type": "uint256",
                "value": 1582225200000
            },
            {
                "type": "address",
                "value": "0x6B175474E89094C44Da98b954EedeAC495271d0F"
            },
            {
                "type": "uint256",
                "value": 20
            },
            {
                "type": "uint256",
                "value": 16
            },
            {
                "type": "uint256",
                "value": 1
            },
            {
                "type": "uint256[]",
                "value": [
                    {
                        "type": "uint256",
                        "value": 50
                    },
                    {
                        "type": "uint256",
                        "value": 30
                    },
                    {
                        "type": "uint256",
                        "value": 10
                    },
                    {
                        "type": "uint256",
                        "value": 10
                    },
                    {
                        "type": "uint256",
                        "value": 0
                    },
                    {
                        "type": "uint256",
                        "value": 0
                    },
                    {
                        "type": "uint256",
                        "value": 0
                    },
                    {
                        "type": "uint256",
                        "value": 0
                    },
                    {
                        "type": "uint256",
                        "value": 0
                    },
                    {
                        "type": "uint256",
                        "value": 0
                    },
                    {
                        "type": "uint256",
                        "value": 0
                    },
                    {
                        "type": "uint256",
                        "value": 0
                    },
                    {
                        "type": "uint256",
                        "value": 0
                    },
                    {
                        "type": "uint256",
                        "value": 0
                    },
                    {
                        "type": "uint256",
                        "value": 0
                    },
                    {
                        "type": "uint256",
                        "value": 0
                    }
                ]
            },
            {
                "type": "string",
                "value": "default"
            },
            {
                "type": "bool",
                "value": 1
            }
        ],
        "transactionHash": "0x719dc02d633a4693ebffde7fbf29acbad98207a3430f6e4234d000a390f11a45",
        "logIndex": 39,
        "blockNumber": 9515193,
        "blockHash": "0x84ad721ff827e0e0abb4e55d5dac39631634e777fa90b30a75d7627004163501",
        "address": "0x2b79f21B5a90aB2204cDb8163708F2077285329c",
        "status": "CONFIRMED",
        "eventSpecificationSignature": "0x4e562d2cea5240426a17df20539e8169f3a02352ebee712ecc3ee340fdebc09a",
        "networkName": "default",
        "id": "0x719dc02d633a4693ebffde7fbf29acbad98207a3430f6e4234d000a390f11a45-0x84ad721ff827e0e0abb4e55d5dac39631634e777fa90b30a75d7627004163501-39"
    },
    "retries": 0
}
```

**Comment by qwwdfsad** [comment]
For this input, the test fails on each platform with `NumberFormatException`, `1582225200000` literal is too large for `Int`. If I fix it, test start to pass.

**Comment by hshar7** [comment]
@qwwdfsad the problem I was having was because strictMode=off, did you remove that option? I don't think strictMode should be off to parse this

**Comment by pdvrieze** [comment]
@hshar7 If the integer value is out of range (and it is for a 32bit int) then strict parsing should treat that as an invalid input.

**Comment by qwwdfsad** [comment]
Figured it out, sorry for the delay.
Yes, it (was) an expected behaviour: your `EventDetails` class declares the only property, `nonIndexedParameters`, while the input JSON contains a lot of "unknown" ones: `indexedParameters`, `nodeName`, etc. Strict mode fairly has thrown an exception when encountering unknown keys in the input data.

But since the next release (this week, hopefully), we reworked it a bit and now there is no `strictMode` property at all. It was split into multiple configuration flags: `ignoreUnknownKeys`, `isLenient` and `serializeSpecialFloatingPointValues`. It will be enough for your use-case to enable `ignoreUnkownKeys`. Moreover, the default exception will suggest that.
